hsmd/hsmd.c

```diff
@@ -1512,87 +1512,63 @@ static void hsm_key_for_utxo(struct privkey *privkey, struct pubkey *pubkey,
 	}
 }
 
-static void sign_input(struct bitcoin_tx *tx, struct utxo *in,
-		       struct pubkey *inkey,
-		       struct bitcoin_signature *sig,
-		       int index)
+/* Find our inputs by the pubkey associated with the inputs, and
+ * add a partial sig for each */
+static void sign_our_inputs(struct utxo **utxos, struct wally_psbt *psbt)
 {
-	struct privkey inprivkey;
-	u8 *subscript, *wscript, *script;
-
-	/* Figure out keys to spend this. */
-	hsm_key_for_utxo(&inprivkey, inkey, in);
-
-	/* It's either a p2wpkh or p2sh (we support that so people from
-	 * the last bitcoin era can put funds into the wallet) */
-	wscript = p2wpkh_scriptcode(tmpctx, inkey);
-	if (in->is_p2sh) {
-		/* For P2SH-wrapped Segwit, the (implied) redeemScript
-		 * is defined in BIP141 */
-		subscript = bitcoin_redeem_p2sh_p2wpkh(tmpctx, inkey);
-		script = bitcoin_scriptsig_p2sh_p2wpkh(tx, inkey);
-		bitcoin_tx_input_set_script(tx, index, script);
-	} else {
-		/* Pure segwit uses an empty inputScript; NULL has
-		 * tal_count() == 0, so it works great here. */
-		subscript = NULL;
-		bitcoin_tx_input_set_script(tx, index, NULL);
-	}
-	/* This is the core crypto magic. */
-	sign_tx_input(tx, index, subscript, wscript, &inprivkey, inkey,
-		      SIGHASH_ALL, sig);
-
-	/* The witness is [sig] [key] */
-	bitcoin_tx_input_set_witness(
-		tx, index, take(bitcoin_witness_p2wpkh(tx, sig, inkey)));
-}
-
-/* This completes the tx by filling in the input scripts with signatures. */
-static void sign_all_inputs(struct bitcoin_tx *tx, struct utxo **utxos)
-{
-	/*~ Deep in my mind there's a continuous battle: should arrays be
-	 * named as singular or plural?  Is consistency the sign of a weak
-	 * mind?
-	 *
-	 * ZmnSCPxj answers thusly: One must make peace with the fact, that
-	 * the array itself is singular, yet its contents are plural. Do you
-	 * name the array, or do you name its contents? Is the array itself
-	 * the thing and the whole of the thing, or is it its contents that
-	 * define what it is?
-	 *
-	 *... I'm not sure that helps! */
-	assert(tx->wtx->num_inputs == tal_count(utxos));
 	for (size_t i = 0; i < tal_count(utxos); i++) {
-		struct pubkey inkey;
-		struct bitcoin_signature sig;
+		struct utxo *utxo = utxos[i];
+		for (size_t j = 0; j < psbt->num_inputs; j++) {
+			struct privkey privkey;
+			struct pubkey pubkey;
+
+			if (!wally_tx_input_spends(&psbt->tx->inputs[j],
+						   &utxo->txid, utxo->outnum))
+				continue;
```

Comment:
Am I correct in assuming this is the case where the UTXO is not under the control of the node? I.e., this output doesn't belong to us and hence we skip it. So this is the case that is alluded to in the commit message?

So this `continue` has two reasons: either we are pairing up the UTXO with the wrong input, or the caller didn't provide the UTXO because it is not under the control of this node.

Reply:
Yes your assumption is correct. We only have `utxo` data for internal-wallet utxos -- which is where these come from.

This check is looking for our inputs on this transaction. Note that we could skip the check/iteration and just call 'straight' `psbt_sign` except that we have to add the public key information to inputs (we could do this elsewhere except for inputs that originated from a close transaction).